torch/csrc/Module.cpp

```diff
@@ -2716,3 +2727,185 @@ struct call_duplicate_guard {
+    py::gil_scoped_acquire gil;
+    underlying_python_guard_.attr("exchange_device")(device.index());
+    return getDevice();
+  }
+  c10::Device getDevice() const override {
+    py::gil_scoped_acquire _;
+    int device_id;
+    if (underlying_python_guard_.is_none()) {
+      std::cout << "I am None" << std::endl;
+      return c10::Device(type(), 0);
+    }
+    py::object py_device_id = underlying_python_guard_.attr("get_device_id");
+    if (py_device_id.is_none()) {
+      device_id = 0;
+    } else {
+      device_id = py_device_id().cast<c10::DeviceIndex>();
+    }
+    return c10::Device(type(), device_id);
+  }
+  void setDevice(c10::Device device) const override {
+    py::gil_scoped_acquire gil;
+    underlying_python_guard_.attr("set_device")(device.index());
+    // no-op
+  }
+  void uncheckedSetDevice(c10::Device device) const noexcept override {
+    py::gil_scoped_acquire gil;
+    try {
+      underlying_python_guard_.attr("set_device")(device.index());
+    } catch (const py::error_already_set& e) {
+    // no-op
+    }
+  }
+  // TODO
+  c10::Stream getStream(c10::Device) const noexcept override {
+    // no-op
+    return c10::Stream(c10::Stream::DEFAULT, getDevice());
+  }
+
+  c10::Stream getNewStream(c10::Device, int priority = 0) const override {
+    // no-op
+    (void)priority;
+    return c10::Stream(c10::Stream::DEFAULT, getDevice());
+  }
+
+  c10::Stream exchangeStream(c10::Stream) const noexcept override {
+    // no-op
+    return c10::Stream(c10::Stream::DEFAULT, getDevice());
+  }
+  c10::DeviceIndex deviceCount() const noexcept override {
+    return 1;
+  }
+
+  // TODO(qihqi): support Event-related functions
+  void record(
+      void** /*event*/,
+      const c10::Stream& /*stream*/,
+      const c10::DeviceIndex /*device_index*/,
+      const c10::EventFlag /*flag*/) const override {
+  }
+  void block(void* /*event*/, const c10::Stream& /*stream*/) const override {
+  }
+  bool queryEvent(void* /*event*/) const override {
+    return true;
+  }
+  void destroyEvent(void* /*event*/, const c10::DeviceIndex /*device_index*/)
+      const noexcept override {}
+
+  // Stream-related functions
+  bool queryStream(const c10::Stream& /*stream*/) const override {
+    return true;
+  }
+  void synchronizeStream(const c10::Stream& /*stream*/) const override { }
+
+  py::object underlying_python_guard_;
+};
+
+bool registerPythonPrivateUse1Hook(py::object hook) {
+  if (at::isPrivateUse1HooksRegistered()) {
+    return false;
+  }
+  at::RegisterPrivateUse1HooksInterface(new PythonPrivateUse1HooksInterface(hook));
+  return true;
+}
+
+bool registerPythonPrivateUse1DeviceGuard(py::object guard) {
+  if (c10::impl::hasDeviceGuardImpl(c10::DeviceType::PrivateUse1)) {
+    return false;
+  }
+  c10::impl::registerDeviceGuard(
+    c10::DeviceType::PrivateUse1,
+    new PythonPrivateUse1DeviceGuard(guard));
+  return true;
+}
+
+at::Tensor createEmptyTensor(
```

Comment:
Why do we need this? This doesn't make sense, doesn't `torch._C._make_wrapper_subclass` or `torch._C._make_subclass` do what you need?